tools/cpp_build/src/import.cpp

```diff
@@ -0,0 +1,462 @@
+				n->ss_(attr->name, std::move(reinterpret_cast<StringsAttr*>(attr)->value()));
+  			break;
+      case AttributeKind::ts:
+				n->ts_(attr->name, std::move(reinterpret_cast<TensorsAttr*>(attr)->value()));
+  			break;
+      // case AttributeKind::gs:
+  		//	 break;
+  	}
+		delete attr;
+		attrs[i] = nullptr;
+  }
+
+  graph->appendNode(n);
+
+  return true;
+}
+
+bool decode_input(pb_istream_t *stream, const pb_field_t *field, void **arg) {
+
+  onnx_ValueInfoProto proto = onnx_ValueInfoProto_init_default;
+
+  auto graph_wrapper = (GraphWrapper*)(*arg);
+  auto graph = graph_wrapper->graph;
+
+  std::string name;
+  proto.name.funcs.decode = &read_string;
+  proto.name.arg = &name;
+
+  if (!pb_decode(stream, onnx_ValueInfoProto_fields, &proto)) {
+    return false;
+  }
+
+  Value* v = graph->addInput();
+  graph_wrapper->value_map[name] = v;
+  auto nodes = graph_wrapper->node_input_map[name];
+
+  for (int i=0; i<nodes.size(); i++) {
+    nodes[i]->addInput(v);
+  }
+
+  return true;
+}
+
+bool decode_output(pb_istream_t *stream, const pb_field_t *field, void **arg) {
+
+  onnx_ValueInfoProto proto = onnx_ValueInfoProto_init_default;
+
+  auto graph_wrapper = (GraphWrapper*)(*arg);
+  auto graph = graph_wrapper->graph;
+
+  std::string name;
+  proto.name.funcs.decode = &read_string;
+  proto.name.arg = &name;
+
+  if (!pb_decode(stream, onnx_ValueInfoProto_fields, &proto)) {
+    return false;
+  }
+
+  if (graph_wrapper->value_map.find(name) != graph_wrapper->value_map.end()) {
+    Value* v = graph_wrapper->value_map.at(name);
+    graph->block()->registerOutput(v);
+  }
+
+  return true;
+}
+
+bool decode_graph(pb_istream_t *stream, const pb_field_t *field, void **arg) {
+
+  onnx_GraphProto proto = onnx_GraphProto_init_default;
+
+  auto graph_wrapper = (GraphWrapper*)(*arg);
+  auto graph = graph_wrapper->graph;
+  auto initializers = graph_wrapper->initializers;
+
+  proto.input.funcs.decode = &decode_input;
+  proto.input.arg = graph_wrapper;
+
+  proto.output.funcs.decode = &decode_output;
+  proto.output.arg = graph_wrapper;
+
+  proto.node.funcs.decode = &decode_node;
+  proto.node.arg = graph_wrapper;
+
+  proto.initializer.funcs.decode = &decode_tensors;
+  proto.initializer.arg = initializers;
+
+  if (!pb_decode(stream, onnx_GraphProto_fields, &proto)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool decode_model(pb_istream_t *stream, GraphWrapper* graph_wrapper) {
+
+  onnx_ModelProto proto = onnx_ModelProto_init_default;
+
+  proto.graph.funcs.decode = &decode_graph;
+  proto.graph.arg = graph_wrapper;
+
+  if (!pb_decode(stream, onnx_ModelProto_fields, &proto)) {
+    return false;
+  }
+
+  return true;
+}
+
+}
+
+void ImportGraph(std::string& serialized_graph,
+                 std::shared_ptr<Graph>& graph,
+                 std::vector<at::Tensor>& initializers) {
+
+  pb_istream_t istream = pb_istream_from_buffer(reinterpret_cast<pb_byte_t *>(&serialized_graph[0]), serialized_graph.size());
```

Comment:
Why can't we use `serialized_graph.data()`?

Reply:
Done